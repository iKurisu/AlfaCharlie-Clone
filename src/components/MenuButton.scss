.menu-button {
  width: 60px;
  height: 60px;
  padding: 20px;
  position: fixed;
  top: 20px;
  right: 33px;
  cursor: pointer;
}

.menu-button .menu-hamburger {
  opacity: 1;
}

.menu-button .menu-cross {
  opacity: 0;
}

.menu-button.open .menu-hamburger {
  opacity: 0;
}

.menu-button.open .menu-cross {
  opacity: 1;
}
