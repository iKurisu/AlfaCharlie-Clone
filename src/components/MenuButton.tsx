import React from "react";
import { Dispatch } from "redux";
import { connect } from "react-redux";
import { AppState } from "store";
import { menuActions } from "modules/menu";
import Hamburger from "./menu_button/Hamburger";
import Cross from "./menu_button/Cross";
import { classList } from "utils/class";
import "./MenuButton.scss";

interface MappedState {
  toggled: boolean;
  onTransition: boolean;
}

interface MappedActions {
  toggle: (toggled: boolean) => () => void;
}

type Props = MappedState & MappedActions;

export const MenuButton = ({
  toggled,
  onTransition,
  toggle
}: Props): JSX.Element => (
  <div
    className={classList([
      "menu-button",
      { open: toggled, disabled: onTransition }
    ])}
    onClick={toggle(toggled)}
  >
    <Hamburger />
    <Cross />
  </div>
);

const mapState = ({ menu }: AppState): MappedState => ({
  toggled: menu.toggled,
  onTransition: menu.onTransition
});

const mapDispatch = (dispatch: Dispatch): MappedActions => ({
  toggle: toggled => () => {
    dispatch(menuActions.toggleMenu());
    setTimeout(
      () => dispatch(menuActions.endTransition()),
      toggled ? 1300 : 2300
    );
  }
});

export default connect(mapState, mapDispatch)(MenuButton);
