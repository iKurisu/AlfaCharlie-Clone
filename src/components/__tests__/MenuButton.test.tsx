import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import { MenuButton } from "../MenuButton";

describe("menu button", (): void => {
  it("renders correctly", (): void => {
    const toggle = jest.fn();
    const button = renderer.create(
      <MenuButton toggled={false} toggle={toggle} />
    );

    expect(button).toMatchSnapshot();
  });

  it("calls the event handler", (): void => {
    const toggle = jest.fn();
    const button = shallow(<MenuButton toggled={false} toggle={toggle} />);

    button.find("div").simulate("click");

    expect(toggle.mock.calls.length).toBe(1);
  });

  it("has correct class", (): void => {
    const toggle = jest.fn();
    const button = shallow(<MenuButton toggled={false} toggle={toggle} />);

    expect(button.find(".menu-button").hasClass("open")).toBe(false);

    button.setProps({ toggled: true });

    expect(button.find(".menu-button").hasClass("open")).toBe(true);
  });
});
