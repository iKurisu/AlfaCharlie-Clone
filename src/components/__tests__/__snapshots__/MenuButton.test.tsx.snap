// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`menu button renders correctly 1`] = `
<div
  className="menu-button"
  onClick={[MockFunction]}
>
  <svg
    className="menu-hamburger"
    viewBox="0 0 72 72"
    x="0"
    y="0"
  >
    <line
      x1="2"
      x2="69"
      y1="9"
      y2="9"
    />
    <line
      x1="2"
      x2="69"
      y1="35"
      y2="35"
    />
    <line
      x1="2"
      x2="69"
      y1="61"
      y2="61"
    />
  </svg>
  <svg
    className="menu-cross"
    viewBox="0 0 59 59"
    x="0px"
    y="0px"
  >
    <line
      x1="6.7"
      x2="53"
      y1="6.5"
      y2="53"
    />
    <line
      x1="6.7"
      x2="53"
      y1="53"
      y2="6.5"
    />
  </svg>
</div>
`;
